File: 백준/Gold/9466. 텀 프로젝트/텀 프로젝트.py
import sys
sys.setrecursionlimit(10**5)

t = int(input())

def solution(n, students):
    global cycle_people, cycle_cnt
    # 모든 팀원이 서로 서로 지목하여 하나의 사이클이 만들어져야함!!
    # 4 -> 7 -> 6 -> 4 : 하나의 사이클 발생!!
    # 3 -> 3 : 자기 자신을 선택하는 것도 하나의 사이클이 된다!
    
    # 단순 방문으로만 체크하면 사이클 판단 불가능
    # 방문을 좀 더 자세하게 나눌 필요가 있어보임!
    # 유형1. 탐색시작 
    # 유형2. 탐색중
    # 유형3. 탐색완료
    # 4(방문) -> 7 -> 6 -> 4(재방문)
    
    # out degree 1인 그래프..!! > 모든 사이클을 찾을 수는 없지만 사이클의 유무 판단
    
    adj = [[] for _ in range(n + 1)]
    for i in range(1, n + 1):
        adj[i].append(students[i - 1])
    
    # 탐색시작(0) / 탐색중(1) / 탐색완료(2)
    visited = [0 for _ in range(n + 1)]
    step= []
    cycle_people = 0
    cycle_cnt = 0
    def dfs(num):
        global cycle_people, cycle_cnt
        # 탐색 중
        visited[num] = 1
        # step 기록 - 백트래킹처럼
        step.append(num)
        for x in adj[num]:
            if visited[x] == 0:
                dfs(x)
            # 탐색중 + 탐색중 = 사이클 발견!!!! 원래는 사이클 발견만 하는 함수!
            elif visited[x] == 1:
                cycle_cnt += 1 
                idx = step.index(x)
                # 사이클에 포함된 사람의 수([4(*), 7, 6])
                cycle_people += len(step) - idx 
        # 탐색 완료
        visited[num] = 2
        # step 기록 - 백트래킹처럼
        step.pop()
    
    for num in range(1, n + 1):
        # 탐색시작
        if visited[num] == 0:
            dfs(num)
            
    return n - cycle_people
        
for _ in range(t):
    n = int(input())
    students = list(map(int, input().split()))
    print(solution(n, students))

File: 백준/Gold/9466. 텀 프로젝트/텀 프로젝트.java
import java.io.*;
import java.util.*;

class Solution {
    // 팀 결성 -> union find / (너비, 깊이) 탐색
    // 방향성 존재 -> union find 제거
    int n;
    int[] students;
    int[] visited;
    List<List<Integer>> adj = new ArrayList<>();
    List<List<Integer>> cycles = new ArrayList<>();
    List<Integer> path = new ArrayList<>();

    public int solution(int n, int[] students) {
        this.n = n;
        this.students = students;
        this.visited = new int[n + 1];
        // adj 초기화
        for (int i = 0; i < n + 1; i++) {
            adj.add(new ArrayList<>());
            ;
        }
        // 인접 리스트 구성
        for (int i = 1; i < n + 1; i++) {
            adj.get(i).add(students[i]);
        }
        // dfs  순회하며 팀 체크
        for (int cur = 1; cur < n + 1; cur++) {
            if (visited[cur] == 0) {
                dfs(cur);
            }
        }

        int val = n;
        for(List<Integer>  cycle: cycles) {
//            System.out.println(cycle);
            val -= cycle.size();
        }
        return val;
    }

    // 팀 만들기 cur -> nxt 이동하는 과정 중 만날 수 있는 상태
    // visited[nxt] = 0 : 방문하지 않음 -> dfs 진행 해야함
    // visited[nxt] = 1 : 방문중 - dfs 진행 중 ->
    // visited[nxt] = 2 : 탐색 완료한 곳 -> 따라서 dfs 진행할 필요 없음.
    // 1(0 -> 1) -> 2 / 2(0 -> 1) -> 3 / 3(0 -> 1) -5 / 5(0 -> 1) -> 2 / 2('1' => root => 종료 => '2')
    void dfs(int cur) {
        // 탐색시작
        visited[cur] = 1;
        path.add(cur);

        for (int nxt : adj.get(cur)) {
            // 방문한 적이 없을 때는 탐색
            if (visited[nxt] == 0) {
                dfs(nxt);
            }
            // 이미 다 돌았음. 볼 필요 없음.
            if (visited[nxt] == 2) {
                continue;
            }
            // nxt는 사이클의 일부.. path : a -> b -> [nxt -> ... -> cur] -> nxt
            if (visited[nxt] == 1) {
                int idx = path.indexOf(nxt);
                List<Integer> cycle = new ArrayList<>();
                for (int i = idx; i < path.size(); i++) {
                    cycle.add(path.get(i));
                }
                cycles.add(cycle);
            }
        }
        // dfs 탐색 종료
        visited[cur] = 2;
        path.remove(path.size() - 1);
    }
}

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int T = Integer.parseInt(br.readLine());

        for (int tc = 0; tc < T; tc++) {
            int n = Integer.parseInt(br.readLine());
            int[] students = new int[n + 1];

            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int i = 1; i <= n; i++) {
                students[i] = Integer.parseInt(st.nextToken());
            }
            Solution solution = new Solution();
            System.out.println(solution.solution(n, students));
        }
    }
}

